### PR TITLE
[AndersenAgnostic] Add lazy cycle detection as an optional technique for Andersen

### PR DESCRIPTION
**Pull request by haved**
When propagating along the edge a->b, it is because a has something new. If b already has everything a has, that might indicate that there is a path from b to a. Lazy cycle detection looks for such a path using a DFS from b, and unifies everything that is on such a path b->*->a. It tries at most once per edge.

### CODE REVIEW

File: jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp [hunk 1]
```diff
@@ -1398,6 +1403,7 @@ PointerObjectConstraintSet::RunWorklistSolver(WorklistStatistics & statistics)
   if constexpr (EnableOnlineCycleDetection)
   {
     static_assert(!EnableHybridCycleDetection, "OnlineCD can not be combined with HybridCD");
+    static_assert(!EnableLazyCycleDetection, "OnlineCD can not be combined with LazyCD");
   }
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp:1407]: What about the combination of Lazy cycle detection and hybrid cycle detection? This is not checked for, is it?

Comment (phate) [reply jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp:1407]: Or can both be enabled at the same time?

Comment (phate) [reply jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp:1407]: It seem to look like it, looking at the rest of the code...

File: jlm/llvm/opt/alias-analyses/LazyCycleDetection.hpp [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+      return subset;
+    }
+    return std::nullopt;
+  }
+
+  /**
+   * @return the number of DFSs performed to look for cycles
+   */
+  [[nodiscard]] size_t
+  NumCycleDetectionAttempts()
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/LazyCycleDetection.hpp:141]: `NumCycleDetectionAttemps() const noexcept`

File: jlm/llvm/opt/alias-analyses/LazyCycleDetection.hpp [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+  {
+    return NumCycleDetectAttempts_;
+  }
+
+  /**
+   * @return the number of cycles detected by Lazy cycle detection
+   */
+  [[nodiscard]] size_t
+  NumCyclesDetected()
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/LazyCycleDetection.hpp:150]: `NumCyclesDetected() const noexcept`

File: jlm/llvm/opt/alias-analyses/LazyCycleDetection.hpp [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+  {
+    return NumCyclesDetected_;
+  }
+
+  /**
+   * @return the number of unifications made while eliminating found cycles
+   */
+  [[nodiscard]] size_t
+  NumCycleUnifications()
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/LazyCycleDetection.hpp:159]: same here

File: jlm/llvm/opt/alias-analyses/LazyCycleDetection.hpp [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+  // A set of all checked simple edges first -> second, to avoid checking again
+  util::HashSet<std::pair<PointerObjectIndex, PointerObjectIndex>> CheckedEdges_;
+
+  // The dfs stack, which may contain the same node multiple times
+  std::stack<PointerObjectIndex> DfsStack_;
+  // Possible states of nodes during the DFS
+  static constexpr uint8_t NodeStateNotVisited = 0;
+  static constexpr uint8_t NodeStateVisited = 1;
+  static constexpr uint8_t NodeStateLeft = 2;
```
Comment (phate) [inline jlm/llvm/opt/alias-analyses/LazyCycleDetection.hpp:177]: What does the state "NodeStateLeft" mean? Left as in opposite of right, or as in remaining?

File: tests/jlm/llvm/opt/alias-analyses/TestLazyCycleDetection.cpp [hunk 6]
```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2023, 2024 Håvard Krogstie <krogstie.havard@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#include "TestRvsdgs.hpp"
```
Comment (phate) [inline tests/jlm/llvm/opt/alias-analyses/TestLazyCycleDetection.cpp:6]: `#include <TestRvsdgs.hpp>`

File: tests/jlm/llvm/opt/alias-analyses/TestLazyCycleDetection.cpp [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  {
+    (void)set.CreateDummyRegisterPointerObject();
+  }
+
+  // Create a graph that looks like
+  //   --> 1 --> 2 --> 3
+  //  /          |
+  // 0           |
+  //  \          V
+  //   --> 5 --> 4
```
Comment (phate) [inline tests/jlm/llvm/opt/alias-analyses/TestLazyCycleDetection.cpp:35]: Thank you!

**Review by phate — CHANGES_REQUESTED** [review]

File: jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp (hunk 1 above)
Comment (haved) [reply jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp:1407]: Yes, they can be combined. They are also from the same paper (Hardekopf 2007)

**Comment by haved** [comment]
@phate I fixed your comments. I replaced "left" with "popped", as that is much more descriptive.

**Review by phate — APPROVED** [review]

**Pull request merged**